Allow the use of the new Android ndk paths

orion78fr opened the pull request:
Hi,

As said in #345 comments, ndk is now installed in the `ndk` folder and the `ndk-bundle` folder installation is obsolete (as shown in the SDK Manager of Android studio).
![image](https://user-images.githubusercontent.com/8122888/123173493-fbc6a700-d47e-11eb-9528-20b49f30b0e9.png)

The new way allow multiple ndk to be installed in parallel.
If there is only one available, the build script pick this one.
If multiple NDK versions are installed, you can specify one with the `ANDROID_NDK_VERSION` env var.

It still fallbacks to the `ndk-bundle` folder just to maintain old behaviour.

I added a check that the ndk is present too, and panic if absent, as it's needed for the headers.

Feel free to ask for any modifications, I'm clearly not a rust expert.
I manually tested all the code paths on my install, but I don't know where the tests for this are written (as they are not present in #345 PR).

Bromeon commented:
I don't have the Android toolchain installed to test myself, but if anyone else can have a look at these changes (or even test them), that would be highly appreciated!

bors try

Waridley commented:
I'm working on understanding the situation here. It seems this parallel NDK feature has caused *quite* a few headaches across the ecosystem. As far as I can tell so far, this is probably the correct solution, but I've yet to get a build working. Currently running into an issue that apparently this PR should fix: https://github.com/rust-lang/rust/pull/85806
There are tons of issues from other repos that have linked to that one, so you can go down quite a rabbit hole following links from there :laughing: 
But it seems that PR is merged but not live yet? Does it really require a custom build of `std` to use ndk `>=23.x.x` right now, or am I missing something?

orion78fr commented:
I don't think these two issues are linked because my PR predates the release of NDK r23 as you can see on the screenshot I included.

The issue I had was that there are 2 types of NDK installs : "NDK (side by side)" and "NDK (obsolete)" (the first and last line on the screenshot).
The side by side one allows multiple versions of the NDK sources to be installed (in different folders in the `ndk` folder) whereas the obsolete one only allows the lastest version (in the `ndk-bundle` folder).

The issue here is that godot-rust currently only supports the obsolete versions of the ndk packaging (the `ndk-bundle` folder that doesn't allow multiple versions to coexist on your system) : https://github.com/godot-rust/godot-rust/blob/deb9dbfd413f8ac751a2267be010b2089ff26d50/gdnative-sys/build.rs#L95

I tried my best to don't break the existing behaviour by doing this :

- If only one version is available in the side by side version, pick it
- If there are multiple versions side by side : ask the user to pick one with the `ANDROID_NDK_VERSION` env var and panic
- If there are no side by side version but the obsolete version is installed, pick it
- If there are no version, panic instead of silently ignoring the required include folder

As for the r23 version, I don't know what's up with that x)

Bromeon commented:
Any update on this one?

Do you guys think merging this PR improves the situation over current `master`, or is there quite a risk of regressions?

Bromeon commented:
bors try

Bromeon commented:
@orion78fr @Waridley Do you think we should include this in v0.10?

The Android CI is [currently failing](https://github.com/godot-rust/godot-rust/runs/4593556600?check_suite_focus=true), so might need adaptation, too:
```rs
error: failed to run custom build command for `gdnative-sys v0.9.3 (/home/runner/work/godot-rust/godot-rust/gdnative-sys)`

Caused by:
  process didn't exit successfully: `/home/runner/work/godot-rust/godot-rust/target/release/build/gdnative-sys-1a836bb63e053ba9/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at 'multiple ndk versions available : ["/usr/local/lib/android/sdk/ndk/23.1.7779620", "/usr/local/lib/android/sdk/ndk/22.1.7171670", "/usr/local/lib/android/sdk/ndk/21.4.7075529"], chose one using ANDROID_NDK_VERSION', gdnative-sys/build.rs:112:30
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
     1: std::panicking::begin_panic_fmt
               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:460:5
     2: build_script_build::header_binding::add_android_include_paths
     3: build_script_build::header_binding::generate
     4: build_script_build::main
     5: core::ops::function::FnOnce::call_once
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
  ```

Waridley commented:
Part of me feels like it would be nice if this tried to auto-select the latest version instead of failing if there are multiple installed and no `ANDROID_NDK_VERSION` variable set... Though it could be very frustrating if some weird bugs happen due to an unexpected version being used or something like that. Anyone happen to know if there is any precedent in other tools for how to behave when multiple NDK versions are detected?

orion78fr commented:
I don't really know how other tools behave in situations like this, so to avoid any weird version confusion I've set it to fail early as the fix for it is easy (setting the env var).
I'm not really familiar with rust ecosystem, but maybe I can choose the latest automatically and print a message with the used version and how to chose another one, printing something during build shouldn't break anything right ?

I've seen that there are some clippy warnings too, I'm going to fix them too later this day if I've some free time.

orion78fr commented:
So I added the `semver` crate as a build dependency for the `gdnative-sys` to parse and sort the version numbers instead of doing it by hand, I think it's better this way.

Here is an example of output from a build :

```
The following warnings were emitted during compilation:

warning: Multiple android ndk versions have been detected.
warning: You should chose one using ANDROID_NDK_VERSION environment variable to have reproducible builds.
warning: Available versions : ["C:\\Users\\orion\\AppData\\Local\\Android\\Sdk\\ndk\\21.4.7075529", "C:\\Users\\orion\\AppData\\Local\\Android\\Sdk\\ndk\\23.1.7779620"]
warning: Automatically chosen version : 23.1.7779620 (latest)
```

I rebased this branch on latest master too as it's 6 month old

orion78fr commented:
Oops, sorry for the rustfmt pass, I was pretty sure I enabled it in CLion's settings, but it seems I did it on another project by mistake.

Bromeon commented:
No worries! It's a security feature of GitHub that first contributors' CI needs to be explicitly approved. Once a first PR is merged, the four checks will run automatically and immediately.

I'll also run the full CI:
bors try

Waridley commented:
I like this solution :+1: 

My only remaining question is whether we should still go ahead and set the environment variable in CI to get rid of the warning/prevent any unforeseen side effects... and why two versions are getting installed in CI in the first place :laughing:

Bromeon commented on gdnative-sys/build.rs:
```diff
@@ -141,8 +141,8 @@ mod header_binding {
                     println!(
-                        "Automatically chosen version : {} (latest)",
-                        ndk_version.to_string()
+                        "cargo:warning=Automatically chosen version : {} (latest)",
```
Maybe mention the reason why this decision was made?
```rs
"cargo:warning=Multiple Android NDKs found; chosing latest version: {}"
```

(also, no space before colon)

orion78fr replied:
It's already specified in the full output :
```
warning: Multiple android ndk versions have been detected.
warning: You should chose one using ANDROID_NDK_VERSION environment variable to have reproducible builds.
warning: Available versions : ["/usr/local/lib/android/sdk/ndk/23.1.7779620", "/usr/local/lib/android/sdk/ndk/22.1.7171670", "/usr/local/lib/android/sdk/ndk/21.4.7075529"]
warning: Automatically chosen version : 23.1.7779620 (latest)
```

Sorry for the spaces before colons, it's because of French typography rules that wants non-breaking spaces before certain punctuation and I'm used to it.

Bromeon replied:
Ah, missed that. Haha, no worries, happens a lot 🙂

Bromeon commented on gdnative-sys/build.rs:
```diff
@@ -91,29 +91,91 @@ mod header_binding {
 
+        let mut android_ndk_root: Option<PathBuf> = None;
+
+        if Path::join(&android_sdk_root, "ndk").exists() {
```
`Path::join(&android_sdk_root, "ndk")` is used again further down, could be extracted into a variable `sdk_path`.

Bromeon commented on gdnative-sys/build.rs:
```diff
@@ -91,29 +91,91 @@ mod header_binding {
+            }
+        }
+
+        if android_ndk_root.is_none() && Path::join(&android_sdk_root, "ndk-bundle").exists() {
```
I'd also extract `Path::join(&android_sdk_root, "ndk-bundle")` to a variable.

It might need to be unnecessarily constructed if the first condition is false, but the code is more readable and the path itself is not constructed twice.

Bromeon commented on gdnative-sys/build.rs:
```diff
@@ -91,29 +91,91 @@ mod header_binding {
+            android_ndk_root = Some(Path::join(&android_sdk_root, "ndk-bundle"));
+        }
+
+        if android_ndk_root.is_none() {
+            panic!("android ndk needs to be installed");
+        };
+
+        let android_ndk_root = android_ndk_root.unwrap();
```
Can be simplified:
```rs
let android_ndk_root = android_ndk_root.expect("Android NDK needs to be installed");
```

orion78fr replied:
That's exactly why I need someone to proofread me for rust, I'm not used to its idioms yet.

Bromeon reviewed (commented):
Thanks for updating this again!
Added a few stylistic comments.

Bromeon commented:
The failing tests were caused by the Rust 1.58 release and have been fixed in #851. Rebasing on latest  `master` should solve it. To check, I'll also run bors, which automatically rebases:

bors try

orion78fr commented:
By the way, I have a weird issue and I don't know if I should include it in this PR or not.

I can't compile on my windows machine because it pulls the jni.h from my java install instead of the one from ndk.
I have AdoptOpenJDK v16.0.1.9 and its jni.h includes a lot of unsupported things for this target :

```
[...]

  C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-hotspot\include/jni.h:1965:1: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions' to enable support for __declspec attributes, err: true
  C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-hotspot\include/jni.h:1965:29: warning: '__stdcall' calling convention is not supported for this target [-Wignored-attributes], err: false
  C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-hotspot\include/jni.h:1969:1: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions' to enable support for __declspec attributes, err: true
  C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-hotspot\include/jni.h:1969:16: warning: '__stdcall' calling convention is not supported for this target [-Wignored-attributes], err: false
  C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-hotspot\include/jni.h:1972:1: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions' to enable support for __declspec attributes, err: true
  C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-hotspot\include/jni.h:1972:16: warning: '__stdcall' calling convention is not supported for this target [-Wignored-attributes], err: false
  thread 'main' panicked at 'Unable to generate bindings: ()', C:\Users\orion\.cargo\git\checkouts\godot-rust-b0f74199a5354947\72e09ea\gdnative-sys\build.rs:270:43
```

I don't think it makes any sense to get jni.h from your native install rather than from the ndk itself (that contains all that's needed for compilation).

Bromeon commented:
> I can't compile on my windows machine because it pulls the jni.h from my java install instead of the one from ndk.

What is "it"?
Is there a way to avoid this inside godot-rust? Or does it depend on the user's setup?

Also, in the end, could you combine the commits to a single one? E.g. with `git rebase -i` or `git reset --soft master && git commit` 🙂

orion78fr commented:
What I mean is that the `JAVA_HOME` env var (and adding the host include folder in clang args) is probably useless, as all the required headers for compilation are in the NDK folder (e.g. the `jni.h` is located in the `ndk\21.4.7075529\sysroot\usr\include` folder too).
Here, my host java headers (windows) are interfering with the ndk ones.
Probably by luck, it doesn't have any effect for a Linux host (as Android is a Linux too, it does compile for android target).

Bromeon commented:
Thanks for elaboration! 

What do you suggest? Should we remove `JAVA_HOME` references from the build scripts and only rely on `ANDROID_SDK_ROOT`?

Also, are those environment variables documented somewhere in godot-rust?

orion78fr commented:
> What do you suggest? Should we remove `JAVA_HOME` references from the build scripts and only rely on `ANDROID_SDK_ROOT`?

Yeah probably, it doesn't seem useful to me, but maybe I'm missing something. At least it compiles for me on windows when removing `JAVA_HOME`.

> Also, are those environment variables documented somewhere in godot-rust?

There are mentioned there : https://godot-rust.github.io/book/exporting/android.html#setting-up-environment-variables-for-gdnative-sys

> My only remaining question is whether we should still go ahead and set the environment variable in CI to get rid of the warning/prevent any unforeseen side effects

Would probably be better, seems like godot recommends version 21 : https://docs.godotengine.org/en/stable/getting_started/workflow/export/exporting_for_android.html#download-the-android-sdk

> and why two versions are getting installed in CI in the first place

I read the ci files and what's weird is that an ndk is downloaded and set (`export ANDROID_SDK_ROOT=/opt/ndk/android-ndk-r21d;`) and completely ignored in the next step (the script search under `/usr/local/lib/android/sdk`, which is probably where it's installed on ubuntu).
Seems like exports are not carried to next CI step (well, it wouldn't work if that was the case because there is no `ndk-bundle` or `ndk` folder where it should be looking).

> Also, in the end, could you combine the commits to a single one?

Doesn't github have an option to squash the commits when merging a PR ?

orion78fr commented:
So yeah, seems like it builds fine without the `JAVA_HOME` var (https://github.com/orion78fr/godot-rust/runs/4836172566?check_suite_focus=true), should I include it with this PR ?

Bromeon commented:
> > What do you suggest? Should we remove `JAVA_HOME` references from the build scripts and only rely on `ANDROID_SDK_ROOT`?
> 
> Yeah probably, it doesn't seem useful to me, but maybe I'm missing something. At least it compiles for me on windows when removing `JAVA_HOME`.

One more thought: does the Android toolchain itself look for `JAVA_HOME`? If you remove this env var from your `PATH`, does the setup still work? Not that there's a hidden requirement that we're not aware of.

If that's not the case, it's good for me to remove `JAVA_HOME`. Then you can gladly include these changes in the PR.

---

> > My only remaining question is whether we should still go ahead and set the environment variable in CI to get rid of the warning/prevent any unforeseen side effects
> 
> Would probably be better, seems like godot recommends version 21 : https://docs.godotengine.org/en/stable/getting_started/workflow/export/exporting_for_android.html#download-the-android-sdk

Do we also want to include that in this PR, or do you prefer it to be separate?

---

> 
> > and why two versions are getting installed in CI in the first place
> 
> I read the ci files and what's weird is that an ndk is downloaded and set (`export ANDROID_SDK_ROOT=/opt/ndk/android-ndk-r21d;`) and completely ignored in the next step (the script search under `/usr/local/lib/android/sdk`, which is probably where it's installed on ubuntu). Seems like exports are not carried to next CI step (well, it wouldn't work if that was the case because there is no `ndk-bundle` or `ndk` folder where it should be looking).

The Android CI scripts are rather old, and there have also been breaking changes to GitHub Actions, so it's well possible that some parts of it are now obsolete.

I think you're right that `export` doesn't work; persisting environment variables across steps needs to be done like this:

https://github.com/godot-rust/godot-rust/blob/685db3178f0c8ca9fdd3e30dda9acc4da45d81d2/.github/workflows/doc.yml#L4

Maybe @Waridley also has something to say regarding all the above suggestions 💡 

---

> > Also, in the end, could you combine the commits to a single one?
> 
> Doesn't github have an option to squash the commits when merging a PR ?

We're not using the merge buttons, because we want to have certainty that all CI passes whenever we merge a branch. Instead, we use the bors bot. Unfortunately, so far it only supports squashing for _all_ merged branches, and there are also valid cases which require more commits (such as complex refactorings with unrelated changes).

What we would need is to have a bors squash option per branch: https://github.com/bors-ng/bors-ng/issues/1028
Until then, I unfortunately have to ask contributors to do it 😕 but let me know if you need help with it!

orion78fr commented:
> One more thought: does the Android toolchain itself look for `JAVA_HOME`? If you remove this env var from your `PATH`, does the setup still work? Not that there's a hidden requirement that we're not aware of.
>
> If that's not the case, it's good for me to remove JAVA_HOME. Then you can gladly include these changes in the PR.

I doesn't seem so, at least on my machine.
It's not in `env | grep JAVA` and compiles just fine.

> Do we also want to include that in this PR, or do you prefer it to be separate?

I just added it to the CI env, so it should be good.

> What we would need is to have a bors squash option per branch: bors-ng/bors-ng#1028
> Until then, I unfortunately have to ask contributors to do it 😕 but let me know if you need help with it!

Oh I see. I can totally do it, but do you have preferences for commit message formatting ? Like putting the same name as the PR or having the PR # ?

Bromeon commented:
> I just added it to the CI env, so it should be good.

Ok, great!

> Oh I see. I can totally do it, but do you have preferences for commit message formatting ? Like putting the same name as the PR or having the PR # ?

No no, as long as it's somewhat descriptive, it's fine. Just maybe list the important changes below the message "header" (1st line).

You don't need to include the PR number, bors will do it automatically. Thanks a lot!

Bromeon commented:
Let me know when it's ready to merge!
Any more feedback regarding this, e.g. @Waridley?

bors try

orion78fr commented:
Looks good to me if you don't have any more feedback. Nightly clippy warnings doesn't relate to the parts I've modified.

Waridley commented:
Looks good to me

Bromeon commented:
@orion78fr @Waridley Thanks a lot for this PR and the review!

bors r+

orion78fr commented:
You're welcome !
By the way, do you want me to contribute to update the documentation accordingly ?

Bromeon commented:
If you have the time, that would be appreciated! 👍
